cclient: implement were_addresses_spent_from

howjmay opened the pull request:
implement were_addresses_spent_from


# Test Plan:
CI must pass

howjmay committed:
cclient: implement were_addresses_spent_from

thibault-martinez commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -769,21 +769,63 @@ retcode_t json_store_transactions_serialize_request(
+  }
+
+  cJSON_AddItemToObject(json_root, "command",
+                        cJSON_CreateString("storeTransactions"));
```
`"wereAddressesSpentFrom"`

howjmay replied:
Sorry I will fix it asap.
However I was referred document of Python code here 
iota.readme.io/reference#storetransactions
The Python code on the above link should be correct

thibault-martinez replied:
You're right that's a mistake, we'll fix that, thanks !

thibault-martinez commented on cclient/serialization/tests/test_json.c:
```diff
@@ -580,11 +581,41 @@ void test_serialize_store_transactions(void) {
+  serializer_t serializer;
+  const char* json_text =
+      "{"
+      "\"command\":\"storeTransactions\","
```
`"wereAddressesSpentFrom"`

howjmay committed:
Correct command name

thibault-martinez commented on cclient/response/were_addresses_spent_from.c:
```diff
@@ -0,0 +1,30 @@
+      return *p;
+    }
+  }
+  return '\0';
```
Shouldn't this be `NULL` instead of `'\0'` ?

howjmay replied:
I will fix it with the problem below

thibault-martinez commented on cclient/serialization/tests/test_json.c:
```diff
@@ -580,11 +581,40 @@ void test_serialize_store_transactions(void) {
+  serializer.vtable.were_addresses_spent_from_deserialize_response(
+      &serializer, json_text, res);
+
+  TEST_ASSERT_EQUAL_STRING("true", were_addresses_spent_from_res_at(res, 0));
```
These should be booleans, not strings.
Maybe wait for https://github.com/iotaledger/entangled/pull/221 to be merged so that you can use what he did.

howjmay commented:
@thibault-martinez hi I wonder what 
`int were_addresses_spent_from_res_num(were_addresses_spent_from_res_t* in);` is.
where could I get more info. about this function?

thibault-martinez commented:
Maybe @oopsmonk can answer that

oopsmonk commented on cclient/serialization/tests/test_json.c:
```diff
@@ -580,11 +581,40 @@ void test_serialize_store_transactions(void) {
+  init_json_serializer(&serializer);
+  const char* json_text =
+      "{"
+      "\"states\":[\"true\"]"
```
the `true` is a boolean value in JSON, not a string.

oopsmonk commented on cclient/request/were_addresses_spent_from.c:
```diff
@@ -0,0 +1,23 @@
+
+were_addresses_spent_from_req_t* were_addresses_spent_from_req_new() {
+  were_addresses_spent_from_req_t* addresses = NULL;
+  utarray_new(addresses, &ut_str_icd);
```
1. Could rename `addresses` to `states` since the JSON object doesn't have an address element. 
2. should be `ut_int_icd`. it's a boolean value.

oopsmonk commented:
@HowJMay 
`were_addresses_spent_from_res_num` returns the size of an array in the response.  
see `find_transactions_res_hash_num`

howjmay commented:
Thank you i will fix it after #221 merged

howjmay committed:
Merge branch 'develop' of github.com:iotaledger/entangled into wereAddr

howjmay committed:
correct string in json into boolean

tsvisabo commented on cclient/serialization/tests/test_json.h:
```diff
@@ -84,6 +84,45 @@
 
 #define TEST_RAW_TRYTES2 "RAWTRYTESHASH99999999"
 
+#define WERE_ADDRESSES_SPENT_FROM_TRYTES                                       \
```
I think hashes and transaction's trytes in most cases can be shared between commands, so could you please rename it to TRANSACTION_TRYTES and see that other tests passes with that trytes?

tsvisabo commented on cclient/response/were_addresses_spent_from.c:
```diff
@@ -0,0 +1,34 @@
+      return *p;
+    }
+  }
+  return NULL;
```
can we change this function to return `retcode_t` and accept `bool*` ?

howjmay replied:
I found it's kind strange to return `retcode_t` intstead of `bool`, since the similar function in the other API returns `bool`.
Should I change or all the similar functions in other APIs, or just leaves it in the original way?
thank you

tsvisabo replied:
hey @HowJMay ,
It is because you return NULL as boolean that i say so, otherwise i agree we could not return `retcode`
so any place that has NULL returned s a boolean should return RC_CCLIENT_OUT_OF_BOUND_INDEX or something like that..
but for this PR this place is enough :)
Thanks

howjmay committed:
define macros and change function return type

Define macros for hashes and transaction_tryes

Change the return type of were_addresses_spent_from_res_at()
into recode_t
if the index is out of range then the function will return
RC_CCLIENT_RES_ERROR

oopsmonk commented on cclient/response/were_addresses_spent_from.c:
```diff
@@ -17,16 +17,18 @@ void were_addresses_spent_from_res_free(were_addresses_spent_from_res_t* res) {
-    if (p) {
-      return *p;
+retcode_t were_addresses_spent_from_res_at(were_addresses_spent_from_res_t* res,
+                                           int index, bool** states) {
```
why use `bool** states` instead of `bool* states`?

howjmay replied:
Since I hope even though the user uses an uninitialized bool pointer as the argument of `were_addresses_spent_from_res_at()` I can promise the `were_addresses_spent_from_res_at()`
will return NULL.

howjmay replied:
Should I change it to `bool* states` ?

oopsmonk commented on cclient/serialization/tests/test_json.c:
```diff
@@ -690,15 +538,16 @@ void test_serialize_were_addresses_spent_from(void) {
       "{"
       "\"command\":\"wereAddressesSpentFrom\","
-      "\"trytes\":[\"" WERE_ADDRESSES_SPENT_FROM_TRYTES
+      "\"addresses\":[\"" TRANSACTION_TRYTES_1 "\",\"" TRANSACTION_TRYTES_2
```
the address is 81 trytes in IOTA protocol.

howjmay replied:
sorry for this mistake.
I referred iota.readme.io when I wrote these code, and I just found the python version of API document is wrong. (https://iota.readme.io/reference#storetransactions)

oopsmonk commented on cclient/serialization/tests/test_json.c:
```diff
@@ -713,14 +562,26 @@ void test_deserialize_were_addresses_spent_from(void) {
+
+  ret = were_addresses_spent_from_res_at(res, 2, &states);
+  TEST_ASSERT_EQUAL_INT(RC_CCLIENT_RES_ERROR, ret);
+  TEST_ASSERT_NULL(states);
```
you allocates `states` each time calling `were_addresses_spent_from_res_at` but never freed it.

howjmay replied:
sorry for the mistake

howjmay committed:
change function return type and key error

oopsmonk commented:
@HowJMay 
Thanks for your help, but `wereAddressesSpentFrom()` has been deprecated and will be removed in a later iota.js library, so this PR will not be merged.

tsvisabo commented:
@HowJMay Thanks, sorry about that :(

howjmay commented:
Thank you @tsvisabo  and @oopsmonk  I learned a lot  for your advices ><
